### PR TITLE
[mono] Run tests for Android via /t:Test

### PR DESCRIPTION
**Pull request by EgorBo**
Integrate Android into the "Test" workflow, e.g.:
```
./build.sh -os Android -arch arm64 -subset Mono+Libs

cd src/libraries/System.Buffers/tests

dotnet build /p:TargetOS=Android /p:TargetArchitecture=arm64 /t:Test
```
^ should run tests for any test suite on an active device or emulator.

Blockers:
1) Test runner doesn't ignore [ActiveIssue]
2) We don't build `libSystem.Security.Cryptography.Native.OpenSsl` for Android, it's used in many places and many test suites crash with DllNotFoundException. I think we should use something like this: https://stackoverflow.com/questions/54656695/how-to-cross-compile-openssl-for-64-bit-android-on-osx-darwin-using-ndk-19
3) Xharness tool is windows-only for Android at the moment.

### CODE REVIEW

**Comment by steveisok** [comment]
What I'd like us to do next is run tests with OpenSSL put in place.  That'll help us see where the gaps are in android.  

@akoeplinger has an android build with OpenSSL included.

File: src/mono/msbuild/AndroidAppBuilder/Templates/AndroidManifest.xml [hunk 1]
```diff
@@ -5,6 +5,8 @@
           a:versionName="1.0">
   <uses-sdk a:minSdkVersion="%MinSdkLevel%" />
   <uses-permission a:name="android.permission.INTERNET"/>
+  <uses-permission a:name="android.permission.READ_EXTERNAL_STORAGE"/>
+  <uses-permission a:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```
Comment (akoeplinger) [inline src/mono/msbuild/AndroidAppBuilder/Templates/AndroidManifest.xml:9]: nit: we shouldn't need external storage permission, we can use TMPDIR to store the results.

File: src/mono/msbuild/AndroidAppBuilder/Templates/MonoRunner.java [hunk 2]
```diff
@@ -44,25 +45,22 @@ public void onStart() {
...
-        WriteLineToInstrumentation("[Mono] Main() returned " + retcode);
-        runOnMainSync (new Runnable() {
+        int retcode = initRuntime(filesDir, cacheDir, docsDir);
+        runOnMainSync(new Runnable() {
             public void run() {
-                finish (retcode, null);
+                Bundle result = new Bundle();
+                result.putInt("return-code", retcode);
+                // Xharness cli expects "test-results-path" with test results
+                result.putString("test-results-path", docsDir + "/testResults.xml");
```
Comment (akoeplinger) [inline src/mono/msbuild/AndroidAppBuilder/Templates/MonoRunner.java:59]: nit: do a File.Exists check to handle cases where the test runner crashes etc. so the xharness client can detect this condition and not try to fetch a non-existing file.

**Review by akoeplinger — APPROVED** [review]

**Comment by akoeplinger** [comment]
AzDO is green despite pending status on GitHub: https://dev.azure.com/dnceng/public/_build/results?buildId=624145&view=results

**Comment by ViktorHofer** [comment]
We plan to obsolete the test target in favor of dotnet test for both local development and CI. I think when we are that point we probably should override the VSTest target for the ios and Android scenarios so that dotnet test works for them as well.

File: src/mono/msbuild/AndroidAppBuilder/Templates/AndroidManifest.xml (hunk 1 above)
Comment (EgorBo) [reply src/mono/msbuild/AndroidAppBuilder/Templates/AndroidManifest.xml:9]: I noticed xamarin.android guys use external storage for logs

File: src/mono/msbuild/AndroidAppBuilder/Templates/MonoRunner.java (hunk 2 above)
Comment (EgorBo) [reply src/mono/msbuild/AndroidAppBuilder/Templates/MonoRunner.java:59]: xharness uninstall the app automatically after tests (even if the app crashes).
but yeah, I'll fix it anyway

**Comment by stephentoub** [comment]
@ViktorHofer, @safern, this was merged with a change that's causing various library-related legs to fail, like the enterprise testing leg and the all-libraries leg:
https://github.com/dotnet/runtime/pull/35640/files#diff-d41f7fd020dc2dd768ef0e7ededfbe70R102
It looks like we didn't run any of those as part of this PR.  Should we be running additional legs for changes that impact testing infrastructure like this?
cc: @jaredpar

**Comment by ViktorHofer** [comment]
This also fails locally (https://github.com/dotnet/runtime/issues/35664#issuecomment-622397784). My guess is that because we aren't building tests in the runtime-live-build pipeline this wasn't caught. I would recommend adding building the tests as well in that pipeline.

**Comment by safern** [comment]
> I would recommend running a very small set of tests as well in that pipeline.

Since the enterprise pipeline is already exercising building and running the tests locally in the machine, wouldn't it make sense to add a trigger when the eng/testing/* folder is changed? 

It takes around 35mins and it seems to be pretty stable overall. 

https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/enterprise/linux.yml#L12-L19

**Pull request merged**